go/vt/mysqlctl/binlogs_gtid.go

```diff
@@ -94,13 +94,21 @@ func ChooseBinlogsForIncrementalBackup(
 			// know this when we look into the _next_ binlog file's Previous-GTIDs.
 			continue
 		}
+		// Got here? This means backupFromGTIDSet does not full contain the current binlog's Previous-GTIDs.
+		// In other words, Previoud-GTIDs have entries on top of backupFromGTIDSet. Which suggests that these
+		// entries were added by the previous binary log.
 		if i == 0 {
+			// Ummm... there _is no_ previous binary log.
 			return nil, "", "", vterrors.Errorf(vtrpc.Code_FAILED_PRECONDITION, "Required entries have been purged. Oldest binary log %v expects entries not found in backup pos. Expected pos=%v", binlog, previousGTIDsPos)
 		}
-		if !prevGTIDsUnion.Union(purgedGTIDSet).Contains(backupFromGTIDSet) {
+		// The other thing to validate, is that we can't allow a situation where the backup-GTIDs have entries not covered
+		// by our binary log's Previous-GTIDs (padded with purged GTIDs). Because that means we can't possibly restore to
+		// such position.
+		prevGTIDsUnionPurged := prevGTIDsUnion.Union(purgedGTIDSet)
+		if !prevGTIDsUnionPurged.Contains(backupFromGTIDSet) {
```

Comment:
I'm sorry if this is a stupid question, but I don't see what changed in the code. According to the description these line changes are the central change in the PR, but isn't the conditional check the same as before. Aren't we just storing the intermediate value in `prevGTIDsUnionPurged` to print later in the logs? There doesn't seem like there is any change to the logic?

Reply:
You're absolutely right and I must have dozed off when writing the original comment. I quoted the wrong change. The relevant one liner change is:
```go
	// The backup position is the GTISset of the last binary log (taken from Previous-GTIDs of the one-next binary log), and we
	// also include gtid_purged ; this complies with the "standard" way MySQL "thinks" about GTIDs: there's gtid_executed, which includes
	// everything that's ever been applied, and a subset of that is gtid_purged, which are the event no longer available in binary logs.
	// When we consider Vitess incremental backups, what's important for us is "what's the GTIDSet that's true when this backup was taken,
	// and which will be true when we restore this backup". The answer to this is the GTIDSet that includes the purged GTIDs.
	// It's also nice for icnremental backups that are taken on _other_ tablets, so that they don't need to understand what exactly was purged
	// on _this_ tablet. They don't care, all they want to know is "what GTIDSet can we get from this".
	incrementalBackupToPosition.GTIDSet = incrementalBackupToPosition.GTIDSet.Union(gtidPurged.GTIDSet)
```

Reply:
Edited the original comment.